File: ScalaPatterns/src/main/scala/behavioral/StrategyPattern.scala
package behavioral

/**
 * Because Scala offers first-class functions, we can use them to express the same concept directly
 */
object StrategyPattern {

  type Strategy = (Int, Int) => Int

  class Context(computer: Strategy) {
    def use(a: Int, b: Int) = println("The result of operation : " + computer(a, b))
  }

  val add: Strategy = _ + _
  val multiply: Strategy = _ * _
  val complexStuff: Strategy = (a, b) => a * b + a - b

  def main(args : Array[String]) {

    new Context(multiply).use(2, 3)
    new Context(add).use(2, 3)
    new Context(complexStuff).use(2, 3)
  }
}

File: JavaPatterns/src/main/java/behavioral/java7/StrategyPattern.java
package behavioral.java7;

/**
 * The strategy pattern defines a family of encapsulated algorithms and lets the algorithm vary independently from clients that use it.
 * The pattern is handy when we need to select an algorithm at runtime.
 *
 * @author Daniel Leon
 */
interface Strategy {
    int compute(int a, int b);
}

class Add implements Strategy {
    public int compute(int a, int b) { return a + b; }
}

class Multiply implements Strategy {
    public int compute(int a, int b) { return a * b; }
}

class ComplexStuff implements Strategy {

    @Override
    public int compute(int a, int b) {
        return a * b + a - b;
    }
}

class Context  {
    private final Strategy strategy;

    public Context(Strategy strategy) { this.strategy = strategy; }

    public void use(int a, int b) {
        System.out.println("The result of operation : " + strategy.compute(a, b));
    }
}

public class StrategyPattern {

    public static void main(String[] args) {

        new Context(new Multiply()).use(2, 3);
        new Context(new Add()).use(2, 3);
        new Context(new ComplexStuff()).use(2, 3);
    }
}
